packages/react-server/core/components/RootElement.js

```diff
@@ -147,3 +148,26 @@ RootElement.scheduleRender = function(element) {
 	var subscribe = callback => updater = callback;
 	return dfd.promise
 }
+
+RootElement.scheduleRender = function(element) {
+	var {listen, when} = (element||{}).props||{};
+	if (!(listen||when)) {
+		return Q(element).then(RootElement.ensureRootElement);
+	}
+
+	// This is what we'll ultimately resolve our return promise with.
+	// It may be changed by the output of `listen` or `when`.
+	var rendered = element;
+
+	// Install the listener right away to start gathering props.
+	// It may be a gated emitter, but we want to make sure we squeeze
+	// props out of it from the beginning if it's not.
+	// Finally gate on the `when`.
+	return Q(listen && RootElement.installListener(element, listen))
+		.then(el => el && (rendered = el))
+		.then(() => when)
+		.then(childProps => childProps
+			?React.cloneElement(rendered, {childProps})
+			:rendered
+		)
```

Comment:
I'm unsure of the complete flow here, but it seems like if `listen` isn't specified on props, `rendered` will never be set, which seems like it means the promise would resolve w/ `undefined`?